File: src/main/scala/com/wordnik/swagger/online/Generator.scala
package com.wordnik.swagger.online

import com.wordnik.swagger.generator.model._
import com.wordnik.swagger.online.configs._
import com.wordnik.swagger.codegen.model._
import com.wordnik.swagger.codegen._
import com.wordnik.swagger.generator.util.ZipUtil;

import org.json4s._
import org.json4s.JsonDSL._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization.{read, write}

import org.apache.commons.io.FileUtils

import com.fasterxml.jackson.databind._
import com.fasterxml.jackson.databind.node._

import java.io.{ File, PrintWriter }

import scala.collection.JavaConverters._

object Generator {
  implicit val formats = SwaggerSerializers.formats("1.2")

  val mapper = new ObjectMapper()
  val factory = mapper.getJsonFactory()

  def clientOptions(language: String) = {
    val clientConfig = language.toLowerCase match {
      case "java" => new JavaClientConfig()
      case "objc" => new ObjCClientConfig()
      case "android" => new AndroidClientConfig()
      case "php" => new PhpClientConfig()
      case "docs" => new StaticDocsConfig()
      case _ => null
    }
    Option(clientConfig) match {
      case Some(c) => c.fields.toList.sortWith(_.getName < _.getName).asJava
      case _ => List().asJava
    }
  }

  def serverOptions(language: String) = {
    val clientConfig = language.toLowerCase match {
      case "jaxrs" => new JaxRSServerConfig()
      case "nodejs" => new NodeJSServerConfig()
      case "restlet" => new RestletServerConfig()
      case _ => null
    }
    Option(clientConfig) match {
      case Some(c) => c.fields.toList.sortWith(_.getName < _.getName).asJava
      case _ => List().asJava
    }
  }

  def generateClient(language: String, input: ClientOptInput) = {
    val opts = input.getOpts()

    // temp dir to write to
    val outputFolder = File.createTempFile("codegen-", "-tmp")
    outputFolder.delete()
    outputFolder.mkdir()
    outputFolder.deleteOnExit()

    new File(outputFolder.getAbsolutePath + File.separator + "files").mkdir()
    val specFolder = outputFolder + File.separator + "specs"
    new File(specFolder).mkdir()

    writeModel(specFolder, input.getModel())

    opts.setOutputDirectory(outputFolder.getAbsolutePath + File.separator + "files")
    opts.properties = (opts.properties.asScala ++ Map("fileMap" -> (specFolder + File.separator + "/api-docs"))).asJava

    val clientConfig = language.toLowerCase match {
      case "java" => new JavaClientConfig()
      case "objc" => new ObjCClientConfig()
      case "android" => new AndroidClientConfig()
      case "php" => new PhpClientConfig()
      case "docs" => new StaticDocsConfig()
      case _ => null
    }

    Option(clientConfig) match {
      case Some(c) => {
        com.wordnik.swagger.codegen.Codegen.templates.clear
        val files = c.generate(opts)
        val zip = new ZipUtil()
        zip.compressFiles(List(
          new File(outputFolder.getAbsolutePath + File.separator + "files"),
          new File(specFolder)).asJava, outputFolder.getAbsolutePath + File.separator + "bundle.zip")
        outputFolder.getAbsolutePath + File.separator + "bundle.zip"        
      }
      case _ => null
    }
  }

  def generateServer(language: String, input: ClientOptInput) = {
    val opts = input.getOpts()

    // temp dir to write to
    val outputFolder = File.createTempFile("codegen-", "-tmp")
    outputFolder.delete()
    outputFolder.mkdir()
    outputFolder.deleteOnExit()

    new File(outputFolder.getAbsolutePath + File.separator + "files").mkdir()
    val specFolder = outputFolder + File.separator + "specs"
    new File(specFolder).mkdir()

    writeModel(specFolder, input.getModel())

    opts.setOutputDirectory(outputFolder.getAbsolutePath + File.separator + "files")
    opts.properties = (opts.properties.asScala ++ Map("fileMap" -> (specFolder + File.separator + "/api-docs"))).asJava

    val clientConfig = language.toLowerCase match {
      case "jaxrs" => new JaxRSServerConfig()
      case "nodejs" => new NodeJSServerConfig()
      case "restlet" => new RestletServerConfig()
      case _ => null
    }

    Option(clientConfig) match {
      case Some(c) => {
        com.wordnik.swagger.codegen.Codegen.templates.clear
        val files = c.generate(opts)
        val zip = new ZipUtil()
        zip.compressFiles(List(
          new File(outputFolder.getAbsolutePath + File.separator + "files"),
          new File(specFolder)).asJava, outputFolder.getAbsolutePath + File.separator + "bundle.zip")
        outputFolder.getAbsolutePath + File.separator + "bundle.zip"        
      }
      case _ => null
    }
  }

  def writeModel(outputFolder: String, model: JsonNode) = {
    val declarationList = model.get("apiDeclarations")

    for(i <- 0 until declarationList.size) yield {
      val json = parse(declarationList.get(i).toString)
      val listing = json.extract[ApiListing]
      val resourcePath = {
        if(listing.resourcePath.startsWith("/")) listing.resourcePath.substring(1)
        else listing.resourcePath
      }
      val filename = outputFolder + File.separator + resourcePath.replace("/", "_")
      val writer = new PrintWriter(new File(filename))
      writer.write(write(listing))
      writer.close()
    }
    model.asInstanceOf[ObjectNode].remove("apiDeclarations")

    val filename = outputFolder + File.separator + "api-docs"
    val writer = new PrintWriter(new File(filename))
    writer.write(model.toString)
    writer.close()
  }
}

File: src/main/java/com/wordnik/swagger/online/Generator.java
package com.wordnik.swagger.online;

import com.wordnik.swagger.codegen.*;
import com.wordnik.swagger.generator.util.*;
import com.wordnik.swagger.codegen.languages.*;

import com.wordnik.swagger.generator.model.*;
import com.wordnik.swagger.util.Json;

import java.util.*;
import java.io.*;

public class Generator {
  public static String generateClient(String language, ClientOptInput codegenInput) {
    if(codegenInput == null)
      return null;

    try {
      if(codegenInput.getOpts() == null)
        codegenInput.setOpts(new ClientOpts());

      File outputDir = getTmpFolder();

      CodegenConfig config = getConfig(language);
      config.setOutputDir(outputDir.getAbsolutePath());

      codegenInput.setConfig(config);

      new Codegen().opts(codegenInput).generate();

      ZipUtil zip = new ZipUtil();
      List<File> fileList = new ArrayList<File>();
      fileList.add(new File(outputDir.getAbsolutePath()));

      zip.compressFiles(fileList, outputDir + File.separator + "bundle.zip");
      return outputDir.getAbsolutePath() + File.separator + "bundle.zip";
    }
    catch (IOException e) {
      e.printStackTrace();
    }
    return null;
  }

  public static List<InputOption> clientOptions(String language) {
    return null;
  }

  public static List<InputOption> serverOptions(String framework) {
    return null;
  }

  public static String generateServer(String framework, ClientOptInput opts) {
    return null;
  }

  static File getTmpFolder() throws IOException {
    File outputFolder = File.createTempFile("codegen-", "-tmp");
    outputFolder.delete();
    outputFolder.mkdir();
    outputFolder.deleteOnExit();

    new File(outputFolder.getAbsolutePath() + File.separator + "files").mkdir();
    String specFolder = outputFolder + File.separator + "specs";
    new File(specFolder).mkdir();
    return outputFolder;
  }

  static CodegenConfig getConfig(String name) {
    if("objc".equals(name))
      return new ObjcClientCodegen();
    else if("java".equals(name)) 
      return new JavaClientCodegen();
    else
      throw new RuntimeException("unsupported client type");
  }
}
